### PR TITLE
Fix encoding/decoding REE Dicts when using streaming IPC

### PR DESCRIPTION
**Pull request by brancz**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-rs/issues/6398

# What changes are included in this PR?

Include dictionaries within REE dicts when recursively flattening all fields of a schema.

# Are there any user-facing changes?

No, just a bug fix.

@alamb @tustvold

### CODE REVIEW

File: arrow-ipc/src/reader/stream.rs [hunk 1]
```diff
@@ -315,4 +315,64 @@ mod tests {
...
+                IpcWriteOptions::default().with_preserve_dict_id(false),
+            )
+            .expect("Failed to create StreamWriter");
+            writer.write(&batch).expect("Failed to write RecordBatch");
+            writer.finish().expect("Failed to finish StreamWriter");
+        }
+
+        let mut decoder = StreamDecoder::new();
+        let buf = &mut Buffer::from(buffer.as_slice());
+        while let Some(_) = decoder
```
Comment (alamb) [inline arrow-ipc/src/reader/stream.rs:367]: I think we should update the test to verify the Batch that comes back in matches the batch that was written -- aka that the data round trips

File: arrow-schema/src/field.rs [hunk 2]
```diff
@@ -375,6 +375,7 @@ impl Field {
             | DataType::FixedSizeList(field, _)
             | DataType::Map(field, _) => field.fields(),
             DataType::Dictionary(_, value_field) => Field::_fields(value_field.as_ref()),
+            DataType::RunEndEncoded(_, field) => field.fields(),
```
Comment (alamb) [inline arrow-schema/src/field.rs:378]: 👍

**Review by alamb — COMMENTED** [review]
Thank you @brancz  -- this makes sense to me

I have one suggestion for the test, but otherwise 🚀

File: arrow-ipc/src/reader/stream.rs (hunk 1 above)
Comment (brancz) [reply arrow-ipc/src/reader/stream.rs]: makes sense, done!

**Comment by brancz** [comment]
Sorry for forgetting to `cargo fmt`, should be fixed now!

**Comment by alamb** [comment]
I am depressed about the large review backlog in this crate. We are looking for more help from the community reviewing PRs -- see https://github.com/apache/arrow-rs/issues/6418 for more

**Review by alamb — APPROVED** [review]
Thank you @brancz

**Comment by alamb** [comment]
🚀

**Pull request merged**